node/pkg/ethereum/polygonimpl.go

```diff
@@ -0,0 +1,240 @@
+		case <-sub.unsubDone:
+		}
+		close(sub.err)
+	})
+}
+
+func (e *PolygonImpl) SubscribeForBlocks(ctx context.Context, sink chan<- *common.NewBlock) (ethereum.Subscription, error) {
+	if e.BaseEth.client == nil {
+		panic("client is not initialized!")
+	}
+	if e.rawClient == nil {
+		panic("rawClient is not initialized!")
+	}
+	if e.rootClient == nil {
+		panic("root client is not initialized!")
+	}
+	if e.rootCaller == nil {
+		panic("root caller is not initialized!")
+	}
+
+	sub := &PolygonPollSubscription{
+		err: make(chan error, 1),
+	}
+
+	opts := &ethBind.CallOpts{Context: ctx}
+	lastConfirmedBlockNum, err := e.rootCaller.GetLastChildBlock(opts)
+	if err != nil {
+		return nil, err
+	}
+
+	e.logger.Info("RCP: initial block", zap.String("eth_network", e.BaseEth.NetworkName), zap.Stringer("block", lastConfirmedBlockNum))
+
+	var BIG_ONE = big.NewInt(1)
+
+	// We would like to generate one block immediately.
+	lastConfirmedBlockNum.Sub(lastConfirmedBlockNum, BIG_ONE)
+
+	timer := time.NewTimer(time.Millisecond) // Start immediately.
+	go func() {
+		for {
+			select {
+			case <-ctx.Done():
+				return
+			case <-timer.C:
+				opts := &ethBind.CallOpts{Context: ctx}
+				newConfirmedBlockNum, err := e.rootCaller.GetLastChildBlock(opts)
+				if err != nil {
+					e.logger.Error("failed to look up latest block", zap.String("eth_network", e.BaseEth.NetworkName), zap.Error(err))
+					sub.err <- fmt.Errorf("failed to read latest block: %w", err)
+					break
+				}
+
+				if newConfirmedBlockNum.Cmp(lastConfirmedBlockNum) > 0 {
+					e.logger.Info("RCP: detected new checkpoint", zap.String("eth_network", e.BaseEth.NetworkName), zap.Stringer("block", newConfirmedBlockNum))
+					blockNum := big.NewInt(0)
+					blockNum.Add(lastConfirmedBlockNum, BIG_ONE)
+					for !(blockNum.Cmp(newConfirmedBlockNum) > 0) {
+						n := big.Int(*blockNum)
+						txHash, err := e.getTxHash(ctx, &n)
+						if err != nil {
+							e.logger.Error("failed to look up tx hash", zap.String("eth_network", e.BaseEth.NetworkName), zap.Error(err))
+							sub.err <- fmt.Errorf("failed to read tx hash: %w", err)
+							break
+						}
+
+						block := &common.NewBlock{
+							Number: &n,
+							Hash:   *txHash,
+						}
+
+						sink <- block
+
+						blockNum.Add(blockNum, BIG_ONE)
+					}
+
+					lastConfirmedBlockNum = newConfirmedBlockNum
+				}
+
+				timer = time.NewTimer(time.Duration(e.DelayInMs) * time.Millisecond)
+			}
+		}
+	}()
+
+	return sub, err
+}
+
+func (e *PolygonImpl) getTxHash(ctx context.Context, number *big.Int) (*ethCommon.Hash, error) {
+	numStr := ethHexUtils.EncodeBig(number)
+	type Marshaller struct {
+		Hash ethCommon.Hash `json:"hash"`
+	}
+
+	var m Marshaller
+	err := e.rawClient.CallContext(ctx, &m, "eth_getBlockByNumber", numStr, false)
+	if err != nil {
+		e.logger.Error("RCP: failed to get block", zap.String("eth_network", e.BaseEth.NetworkName),
+			zap.String("requested_block", numStr), zap.Error(err))
+		return nil, err
+	}
```

Comment:
```suggestion
	if err := e.rawClient.CallContext(ctx, &m, "eth_getBlockByNumber", numStr, false); err != nil {
		e.logger.Error("RCP: failed to get block", zap.String("eth_network", e.BaseEth.NetworkName),
			zap.String("requested_block", numStr), zap.Error(err))
		return nil, err
	}
```

If you want, you can collapse this since you only use `err` inside that conditional.

Reply:
Changed.